feat: support embedded metadata in trace files

DavePearce opened the pull request:
This adds support for embedding metadata into the generated trace file.  The metadata is sourced from two places: the `zkevm.bin` file, and also some information gathered by the tracer (e.g. block range, etc).

DavePearce committed:
support metadata

This adds support for embedding metadata into the generated trace file.

DavePearce committed:
apply spotless

FlorianHuc commented on arithmetization/src/main/java/net/consensys/linea/zktracer/ZkTracer.java:
```diff
@@ -317,4 +337,9 @@ public Map<String, Integer> getModulesLineCount() {
   }
+
+  public static byte[] getMetadataBytes(Map<String, String> metadata) throws IOException {
+    ObjectWriter mapper = new ObjectMapper().writer();
```
This can be made static to avoid reinstantiating it each time.

DavePearce replied:
Agreed, done.

FlorianHuc commented on arithmetization/src/main/java/net/consensys/linea/zktracer/ZkTracer.java:
```diff
@@ -73,6 +75,11 @@ public class ZkTracer implements ConflationAwareOperationTracer {
   @Getter private final List<Exception> tracingExceptions = new FiniteList<>(50);
 
+  // Fields for metadata
+  private long startBlock = Long.MAX_VALUE;
```
The logic to update start and end block is not straightforward to get. Can't we set them when the zktracer is updated? You have all the info when you instantiate the zktracer object in GenerateConflatedTracesV2

DavePearce replied:
Ok, I have resolved this now ... but it took a few changes to thread through the relevant information.

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/ZkTracer.java:
```diff
@@ -73,6 +75,11 @@ public class ZkTracer implements ConflationAwareOperationTracer {
+  // Fields for metadata
+  private long startBlock = Long.MAX_VALUE;
+  private long endBlock = 0;
+  private BigInteger chainId;
```
shouldn't it be final ?

DavePearce replied:
Agreed, done.

letypequividelespoubelles reviewed (approved):
LGTM

DavePearce committed:
address review comments; embed line counts

This addresses the reviewer comments by avoiding the recalculation of
the starting / ending block number.  This also embeds line counts and
updates the tracer to support structured metadata in the LT file.

DavePearce commented:
@letypequividelespoubelles @FlorianHuc I have now updated this PR to additionally embed line counts.  However, there is currently an issue (to some extent) because it doesn't embed line counts for non-tracing modules, such as keccack.  I need to figure out where that information comes from.

DavePearce committed:
apply spotless

letypequividelespoubelles committed:
add linecount of traceless modules in metadata

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>
